src/sync/app.cpp

```diff
@@ -734,6 +732,113 @@ void App::link_user(std::shared_ptr<SyncUser> user,
     
     App::log_in_with_credentials(credentials, user, completion_block);
 }
+    
+    void App::refresh_custom_data(std::shared_ptr<SyncUser> sync_user,
+                                  std::function<void(Optional<AppError>)> completion_block)
+    {
+        refresh_access_token(sync_user, completion_block);
+    }
+    
+    void App::do_authenticated_request(Request request,
+                                       std::shared_ptr<SyncUser> sync_user,
+                                       std::function<void (Response)> completion_block) const
+    {
+        auto handler = [completion_block, request, sync_user, this](const Response& response) {
+            if (auto error = check_for_errors(response)) {
+                App::handle_auth_failure(error.value(), response, request, sync_user, completion_block);
+            } else {
+                completion_block(response);
+            }
+        };
+
+        request.headers = get_request_headers(sync_user,
+                                              request.uses_refresh_token ?
+                                              RequestTokenType::RefreshToken : RequestTokenType::AccessToken);
+        m_config.transport_generator()->send_request_to_server(request, handler);
+    }
+    
+    void App::handle_auth_failure(const AppError& error,
+                                  const Response& response,
+                                  Request request,
+                                  std::shared_ptr<SyncUser> sync_user,
+                                  std::function<void (Response)> completion_block) const
+    {
+        auto transport_generator = m_config.transport_generator();
+        auto access_token_handler = [&transport_generator,
+                                     &request,
+                                     completion_block,
+                                     response,
+                                     sync_user](const Optional<AppError>& error) {
+            if (!error) {
+                // assign the new access_token to the auth header
+                request.headers = get_request_headers(sync_user, RequestTokenType::AccessToken);
+                transport_generator->send_request_to_server(request, completion_block);
+            } else {
+                // pass the error back up the chain
+                completion_block(response);
+            }
+        };
+        
+        // Only handle auth failures
+        if (error.error_code.value() != 401) {
+            completion_block(response);
+            return;
+        }
+        
+        if (request.uses_refresh_token) {
+            if (!sync_user) {
+                sync_user->log_out();
```

Comment:
What is this supposed to do? Dereferencing a null pointer will crash.

Reply:
impl is wrong, changed to `if (sync_user && sync_user->is_logged_in())`